### ISSUE TITLE
Fetch ISession data from client code without refreshing the session

### ISSUE THREAD

**Issue by martelskiy** [aspnet/Session#229]
I need to check whether a session is not expired in my client code. But by polling each few minutes HTTP GET and checking session status I'm extending the actual session. Which means unless the user closes the browser tab session will never be expired. Is there any way to specify which GET requests should not extend the session?

**Comment by Tratcher** [comment]
Hmm. 
https://github.com/aspnet/Session/blob/c0396f7ff4cfa24caa1357fdd9f1ca22ab4ada80/src/Microsoft.AspNetCore.Session/SessionMiddleware.cs#L107-L117

How about setting `context.Session = null;` after you're done checking the session? That should prevent it from being refreshed.

**Comment by martelskiy** [comment]
> Hmm.
> [Session/src/Microsoft.AspNetCore.Session/SessionMiddleware.cs](https://github.com/aspnet/Session/blob/c0396f7ff4cfa24caa1357fdd9f1ca22ab4ada80/src/Microsoft.AspNetCore.Session/SessionMiddleware.cs#L107-L117)
> 
> Lines 107 to 117 in [c0396f7](/aspnet/Session/commit/c0396f7ff4cfa24caa1357fdd9f1ca22ab4ada80)
> 
>      await _next(context); 
>  } 
>  finally 
>  { 
>      context.Features.Set<ISessionFeature>(null); 
>   
>      if (feature.Session != null) 
>      { 
>          try 
>          { 
>              await feature.Session.CommitAsync(context.RequestAborted); 
> How about setting `context.Session = null;` after you're done checking the session? That should prevent it from being refreshed.

Thank you for your response. I will try it out

**Comment by martelskiy** [comment]
> Hmm.
> [Session/src/Microsoft.AspNetCore.Session/SessionMiddleware.cs](https://github.com/aspnet/Session/blob/c0396f7ff4cfa24caa1357fdd9f1ca22ab4ada80/src/Microsoft.AspNetCore.Session/SessionMiddleware.cs#L107-L117)
> 
> Lines 107 to 117 in [c0396f7](/aspnet/Session/commit/c0396f7ff4cfa24caa1357fdd9f1ca22ab4ada80)
> 
>      await _next(context); 
>  } 
>  finally 
>  { 
>      context.Features.Set<ISessionFeature>(null); 
>   
>      if (feature.Session != null) 
>      { 
>          try 
>          { 
>              await feature.Session.CommitAsync(context.RequestAborted); 
> How about setting `context.Session = null;` after you're done checking the session? That should prevent it from being refreshed.

I guess this would work only if to catch request before middleware kicks in. How to do this? Middleware probably the first layer before a request is getting into asp.net controller

**Comment by Tratcher** [comment]
No, that's the cleanup code that runs at the end of a request and saves the session. Any component after the session middleware could null it out to prevent it from being saved. E.g. you would only do this in your status check endpoint.

**Comment by martelskiy** [comment]
That's basically the way I check whether a user is authenticated. And `_httpContextAccessor.HttpContext.Session = null;` doesn't prevent from session being refreshed. This is full piece of code:
```
[ResponseCache(NoStore = true, Duration = 0)]
        [HttpGet("CheckSessionIsValid")]
        public EmailLoginAuthenticationResponse CheckSessionIsValid()
        {
            var emailLogin = _httpContextAccessor.HttpContext.Session.GetString("_emailLogin");
            if (emailLogin == null)
            {
                return null;
            }
            var deserializedEmailLogin = JsonConvert.DeserializeObject<EmailLoginAuthenticationResponse>(emailLogin);

            _httpContextAccessor.HttpContext.Session = null;

            return deserializedEmailLogin;
        }
```

**Comment by Tratcher** [comment]
Are you using a local in-memory store (default) or a distributed store (e.g. redis)? The in memory one implicitly renews the sliding expiration when accessed where the dististributed one requires an extra action when committing the session. It looks like the workaround above would only work for the distributed one.

**Comment by martelskiy** [comment]
Yes, I use in-memory store. So switching to distributed might fix the issue?

**Comment by Tratcher** [comment]
Drat, I reviewed the [distributed cache](https://github.com/aspnet/Caching/blob/9380a26c408d08015b9062db11f3ba3782b66418/src/Microsoft.Extensions.Caching.Redis/RedisCache.cs#L63-L73) and it has an implicit refresh as well.

This is going to require a different design. You can't load the session and avoid refreshing it. You need a different mechanism to seed your timer. 

Back to the original problem. What are you timing anyways? Is there important information stored in the Session that you don't want users to loose like a multi-page form? Or are you actually trying to time out an Authentication session and sign out the user?

**Comment by martelskiy** [comment]
Yes, what I want to achieve is to sync asp.net core session with front end. Just to be able to retrieve session time stamp to be able to redirect or show some sort of notifications about the session is about to expire etc. We use external service for authentication and store authentication data in cookies. And the actual way to check whether a session is still available is to check if there is some data in cookies. I was thinking about SignalR and attaching timestamp in each request header and set timer in client. But probably no need to reinvent the wheel and some mechanism already exist? Sounds like very trivial case

**Comment by Tratcher** [comment]
It's a common topic but I have yet to see any two cases with the same requirements. It really depends on what activity you want to reset your countdown. Also note that auth session should not be in any way associated with browser session, the semantics are completely different. Consider browser sessions an untrusted, lossy cache where an auth session is a trusted identification mechanism. Logout noticies should track the auth session/cookie, not the browser session. I've seen this accomplished by embedding expiry information in pages rather than making additional queries to get it.
